### ISSUE TITLE
[BUG] Adding a 3DS2 card using ThreeDS2Component fails without location access

### ISSUE THREAD

**Issue by rajvy1** [Adyen/adyen-ios#1176]
**Description**
On updating Adyen framework to the latest version `5.1.0`, the ability to add a `3DS2` card no longer works  with location access turned off. The callback from the framework never returns.  However, it works well when the app has location access.

**To Reproduce**
1. Turn off location access of the app, where adyen-ios sdk is integrated
1. Use any 3DS2 card and try to authenticate it using `AdyenActions.ThreeDS2Component`
2. On receiving fingerprint challenge, call `adyenThreeDS2Component.handle(action)`
3. Wait for delegate callback methods `didComplete(:)`, `didProvide(:)` or `didFail(:)` to execute

**Expected behavior**
Delegate methods `didComplete(:)`, `didProvide(:)` or `didFail(:)` are expected to be called by the framework

**Actual behaviour**
Delegate methods are not called.

**Smartphone (please complete the following information):**
 - Device: iPhone11
 - OS: iOS16.3.1
 - SDK Version 5.1.0

**Comment by descorp** [comment]
Hey @rajvy1 

Thanks for feedback!

> Turn off location access of the app, where adyen-ios sdk is integrated

How exactly do you reproduce this step?

Just checking - do you persist `ThreeDS2Component` as local property?

**Comment by rajvy1** [comment]
Hi @descorp!

> How exactly do you reproduce this step?

I'm sorry do you mean, how we turn off the location access? or how we've integrated the framework?

> do you persist ThreeDS2Component as local property?

We do retain the `ThreeDS2Component` in memory (instance variable) until the transaction or challenge is completed with success/error. 
I'd like to mention again that the 3DS2 component works as expected when location access is turned on.

**Comment by descorp** [comment]
Hey @rajvy1 

Sorry, I mean how do you turned on/off location access?

**Comment by rajvy1** [comment]
Hi @descorp Sorry for the late reply,
That just means our app uses Location Access from iOS for various puposes.
We use `CLLocationManager` to request permission; 
Everytime we launch the app, the user get's a prompt to authorize permission (example below). 

In this case, we happened to notice that somehow **denying** that permission (Tapping "Don't Allow") to the app
causes the 3DS2 transaction to fail later during payment, which isn't expected behaviour from the SDK.

<img src="https://user-images.githubusercontent.com/105429948/229825044-48b8b704-5a1c-4ae6-a95b-8c18020370ff.jpg" width="200" height="400">

**Comment by rajvy1** [comment]
@descorp Any updates here?

**Comment by descorp** [comment]
Hey @rajvy1 

Sorry for delay. Somehow your previous message escaped our notice.
We will run local tests and will get back to you!

**Comment by descorp** [comment]
Hey @rajvy1 

Sorry, still no success in reproducing this issue..
Could you create a small "sandbox" app where you reproduced this behaviour?

**Comment by rajvy1** [comment]
Hi @descorp,

I'm trying to create a sandbox app to reproduce an issue, the challenge I'm currently facing is creating a real transaction/token with a ThreeDS2 Challenge and then asking the Adyen SDK to authorise it,
is difficult without having the involvement of our backend.

I'm aware of 3DS2 test cards, but is there a way to simulate a successful 3DS2 authorization without having a BE support? or would you recommend any other approach?

Thanks
Rajat

**Comment by descorp** [comment]
Hey [@rajvy1](https://github.com/rajvy1)

You can connect directly to Adyen API (for test purpose only) the same way as our [Demo App](https://github.com/Adyen/adyen-ios/tree/develop/Demo) do.
As an alternative, you can use our Demo app as a foundation for your experiment - fork it and then share the link or make PR for e.x.

---

**If you decide to share the code publicly - make sure not to expose your API Key and Client Key.**

**Comment by rajvy1** [comment]
Hi @descorp 

Thanks for the suggestion to fork the repo.

In the demo app, I don't see a flow for testing the standalone `AdyenActions.ThreeDS2Component` which is the one we have integrated in our application. So I've added a new row for the same.

Although, this is where I get stuck: after adding the our demo-server-key and client-key, I try to initiate a new transaction this way:
`threeDS2Component.handle(ThreeDS2FingerprintAction(fingerprintToken: mock3DS2FingerprintToken, paymentData: nil))`.

But this errors out with the below payload: 

```
{
  "status" : 422,
  "errorCode" : "14_002",
  "message" : "Required field 'paymentData' is not provided.",
  "pspReference" : "RP2J7WMM7CLZNN82",
  "errorType" : "validation"
}
```
<img width="310" alt="Screenshot 2023-06-20 at 4 04 56 PM" src="https://github.com/Adyen/adyen-ios/assets/105429948/19eb97d1-f949-4b81-9570-720220ea5468">



I had thought the demo-api-key would simulate a positive authentication flow, but the paymentData is obviously missing, as that is provided by BE. I'm not sure on the next step here.
This is the link to a branch in forked repo: https://github.com/rajvy1/adyen-ios/tree/demo-3ds2.

Let me know how to go about this, or if this helps you identify and troubleshoot the original issue about our 3DS2 integration.

Thanks
Rajat

**Comment by descorp** [comment]
Hey @rajvy1

`ThreeDS2Component` is an "Action Handling Component"; "actions" are additional steps/challenges that shopper need to complete to make a payment. 

To handle "action" one need to make a payment :)
You can try "Card" flow with test card `5454 5454 5454 5454` `03/30` `737` to execute native 3DS2 flow.

**Comment by rajvy1** [comment]
Hi @descorp,

Like I'd also mentioned in the original message, unfortunately, 
I'm afraid the point is to reproduce the issue with `ThreeDS2Component`, whether `CardComponent` works or not is not relevant to our use case. 

And If this means it's anyway not possible to sandbox this issue at our end, now, that you understand our use-case;
Could you try again at your end?

Thanks
Rajat

**Comment by descorp** [comment]
Hey @rajvy1 

Your original description is 

```
To Reproduce

1. Turn off location access of the app, where adyen-ios sdk is integrated
2. Use any 3DS2 card and try to authenticate it using AdyenActions.ThreeDS2Component
3. On receiving fingerprint challenge, call adyenThreeDS2Component.handle(action)
4. Wait for delegate callback methods didComplete(:), didProvide(:) or didFail(:) to execute
```

What is not relevant exactly?

**Comment by rajvy1** [comment]
> You can try "Card" flow with test card 5454 5454 5454 5454 03/30 737 to execute native 3DS2 flow.

Does this not mean to use a `CardComponent`?

**Comment by descorp** [comment]
Yes. Or you can use `CSE`:

```swift
let unencryptedCard = Card(number: "5454545454545454", securityCode: "737", expiryMonth: "03", expiryYear: "2030")
let encryptedCard = try CardEncryptor.encrypt(card: unencryptedCard, with: publicKey)
```

How are you authenticating cards?

**Comment by rajvy1** [comment]
Roughly, authentication challenges like `.fingerprint` or `.challenge` are created by our backend and our app passes those to Adyen's `ThreeDS2Component` to handle. We do not use `CardComponent` anywhere in our integration.

**Comment by descorp** [comment]
Hm..

@rajvy1 where do you get and how do you encrypt shoppers' card data?

**Comment by rajvy1** [comment]
@descorp Sorry for the delay,
We have a separate BE API for tokenization of a card.

**Comment by descorp** [comment]
@rajvy1 

Are you PCI compliant?
Are you using [/payments](https://docs.adyen.com/api-explorer/Checkout/68/post/payments) Checkout API?
How do you collect and encrypt shoppers' card data before requesting `/payments` via Checkout API?

**Comment by mohamed-sqaure1** [comment]
I have the same issue and the same use case scenario even if the app has location access, 

Any Visa card that has 3D security and should use the ThreeDS2Component handler isn't fire the delegate functions

Any updates, please?

**Comment by descorp** [comment]
@mohamed-sqaure1 

> I have the same issue and the same use case scenario even if the app has location access,

I don't think it is related.

**Comment by descorp** [comment]
@rajvy1 
Were you able to resolve this issue?

**Comment by rajvy1** [comment]
Hi @descorp 
It was resolved in `5.2.0` at least on the test env. Since, we could not push this version to production due to other issues, I can't confirm the resolution. I would try to upgrade again in the new year with the latest adyen SDK version and will report back.

**Comment by descorp** [comment]
I am going to close this thread.
@rajvy1 feel free to continue discussion here or reopen it.
